### DISCUSSION TITLE
Implementation of heat conduction equation with robin and Neumann boundary conditions

### DISCUSSION THREAD

**Discussion by abhishek-appana** [kinnala/scikit-fem#1086] [Q&A] [answered]
Hello everyone,

I am working to solve the heat conduction equation with Robin and Neumann boundary conditions on different edges of the boundary. I am trying to modify the code given for example 19 to make it happen. I am not able to understand some parts of that code (lines 62 & 67).

What does 'mass' represent on line 62?

What is the function 'penalize' on line 67?

Can someone direct me to any available documentation on these?

Thank you!

**Comment by gdmcbain** [comment, marked as answer]
`mass` is the "mass" linear form which when assembled for a given basis gives the [Gramian](https://en.m.wikipedia.org/wiki/Gram_matrix) of that basis; here `M`. This `M` is usually called the "mass matrix" in the finite element method; e.g. Ern & Guermond (2004, _Theory and Practice of Finite Elements_; &sect;3.3.4). It invariably arises in transient or eigenvalue problems.  Definition:

https://github.com/kinnala/scikit-fem/blob/68bd6fe9978171cc73b2fb23f80962c687b5c64d/skfem/models/poisson.py#L18

`penalize` imposes the essential boundary condition, here zero, by dividing excursions by a very small number to produce a very big residual unless the value on the boundary is almost the desired value. It's an alternative to `enforce` and `condense`. Penalization is discussed by E. & G. in &sect;8.4.4.

**Reply by abhishek-appana** [reply]
Thank you for your response @gdmcbain. I see.

I do not have zero boundary conditions in my geometry. I implement the boundary conditions as terms in the variational formulation as I have done in this example #1084. Does that mean I can ignore the `D=basis.get_dofs()` part in the `penalize` function?

**Reply by kinnala** [reply]
FEM has two types of boundary conditions. One type is "natural" to the given variational formulation and does not require constraining any unknowns in the linear system. Another type is "essential" which involves constraining unknowns to specific values. The functions condense, enforce, penalize, are all used for "essential" boundary conditions. Since you have only natural boundary conditions, none of these functions is are required.

**Reply by abhishek-appana** [reply]
Got it!

Thank you so much!

**Reply by gdmcbain** [reply]
Yes   One way to think about `penalize` is that it approximates an essential (Dirichlet) condition with a natural (Robin) condition using a very high conductance between the prevailing and desired values on the boundary.

**Reply by abhishek-appana** [reply]
I understand. Thank you @gdmcbain.

Probably a dumb question:
Can someone direct me to a reference which explains the math in lines 61-62 & 68-69? I have been reading about Crank Nicolson method from various sources but I am not able to figure out how it is applied here! None of the reading I did, has the test function `v` when applying Crank Nicolson.

From what I understand from the code:
$L = k * \nabla u . \nabla v $
$M = u.v $

How are these 2 expressions derived from the heat equation $\partial u/ \partial t = k \Delta u$? I tried writing the variational formulation using the standard procedures. The formulation I get is:
$\int uvdx + k \delta t \int (\nabla u . \nabla v) dx = \int u^n vdx$
where $u^{n+1}$ is written as $u$ and $u^n$ is the value at timestep $n$. If the first term is written as $M$ and the second term as $L$ (with $\delta t$ multiplied in lines 68-69), what happened to the right hand side?

I also do not understand the mathematical basis of writing $A$ and $B$ from these two!

I am trying to develop this formulation for the below variation of heat equation. Hence the question:
$\rho C_p \partial u/\partial t = k\Delta u + q_{gen}$

Thank you so much for the help, in advance

**Reply by gdmcbain** [reply]
In these expressions for `L` and `M`, we need to expand the trial and test functions `u` and `v` in terms of the basis. (The same basis for a Galerkin method.) Say the elements of the basis are phi[i] then u(x) is the sum over i of a[i] * phi[i](x), where a[i] is the coefficient of u corresponding to the i-th basis function, phi [i].  For a Lagrangian basis, like a lot of the most common finite element bases, the basis functions are one at one node and zero at the others so a[i] = u(x[i]) which we simplify with an abuse of notation as u[i] so that now u is no longer a continuous function of x but a discrete vector of nodal values (or expansion coefficients). Then since the u[i] don't depend on x but the phi [i] do, the form (u, v) expands as (u[j] phi[j], phi[i]) where there is an applied summation over j and where we have put phi [j] for v; v can be anything and this is a convenient choice, it's the Galerkin method. Now the coefficient u[j] comes out of the inner product because it's independent of x: (phi[j], phi[i]) u[j] = M[i, j] * u[j] or in matrix-vector form M@ u.
   That is, the mass and stiffness matrices M and L are formed from the basis functions rather than the test and trial functions.

**Reply by gdmcbain** [reply]
There are a few other steps.  Integrate _&rho;cu_ <sub>,<i>t</i></sub> from _t_ to _t + h_ to get (_&rho;c_) (_u_ &minus;  _u_<sub>old</sub>), assuming _&rho;c_ depends on _x_ but not _t_.  Then the above discretization leads to a mass matrix which includes these coefficients, `M[i, j]` = (_&rho;c&phi;_ <sub><i>j</i></sub>, _&phi;_ <sub><i>i</i></sub>), and the temporal term discretizes to `M @ (u - uold)`.

Then integrate the other term, the conduction term, using say the [trapezoidal rule](https://en.wikipedia.org/wiki/Trapezoidal_rule) and it will look like `h * L @ (u + uold) / 2`,  for an appropriate definition of the  stiffness matrix `L`, which will fall out from the  same kind of expansion of the test  function _u_ in terms of the basis and the application of the Galerkin method.  

Then separate the terms in `u` onto the left-hand  side and those in `uold` onto the right-hand side.  This will leave both `M` and `L` on both sides.  Regroup these to get `A` and `B`.

**Reply by gdmcbain** [reply]
Viz., 

```
M (u - uold) = h L (u + uold) / 2 
(M - h L / 2) u = (M + h L / 2) uold 
```

from which 
```python
A = M + h * L / 2
B = M - h * L / 2
```

**Reply by gdmcbain** [reply]
> Can someone direct me to a reference which explains the math in lines 61-62 & 68-69? 

One reference is &sect;6.1 &lsquo;Parabolic problems&rsquo; of Ern & Guermond  (2004), but it's got a lot of functional analysis and estimation of error which may obscure these basic ideas.

**Reply by abhishek-appana** [reply]
I guess I finally understand it. Thank you so much @gdmcbain. It's really helpful.

Just want to confirm: 
> `M (u - uold) = h L (u + uold) / 2`

Shouldn't this be `M (u - uold) = - h L (u + uold) / 2`?

Below is my math. Please correct me if I'm wrong

$\partial u/ \partial t = k \Delta u $
$\partial u/ \partial t * v= k \Delta u * v$
Integrating on both sides and using integration by parts on the right-hand side, as the test function vanishes at the boundary, we get:
$\int \partial u/ \partial t * v dx= - \int k \nabla u \nabla v dx$
Assuming $u = \Sigma_j \alpha_j(t)\varphi_j(x)$, $v = \varphi_i(x)$ and $M = \int \varphi_j(x) \varphi_i(x) dx$, $L = \int k \nabla \varphi_j \nabla \varphi_i dx$, which makes the above equation:
$M \alpha ' = - L \alpha $
As you have explained above, taking $\alpha(j) = u(j) $, expanding the equation gives:
`M (u - uold) = - h L (u + uold) / 2`
`A u = B u_{old}`

Can you let me know if I am doing something wrong?

Thank you once again!

**Reply by gdmcbain** [reply]
Ha, no, yes, quite right; the sign error is mine.

**Reply by abhishek-appana** [reply]
Okay got it! Thank you!

Just one last question:

When I follow this procedure, my heat equation i.e., $\rho c \partial u / \partial t = k \Delta u + Q_{gen}$ ends up in the form:

$A u = B u_{old} - dt * \int_{edges} \varphi_i q_{loss} ds + dt * \int Q_{gen} \varphi_i dx$ (My test function doesn't vanish on boundaries)

How do I solve this in scikit-fem?

**Reply by gdmcbain** [reply]
The last term looks like a volumetric source.  This is handled as in

https://github.com/kinnala/scikit-fem/blob/6159aa309cd182578405ee8a84553d9a6d7462de/docs/examples/ex01.py#L23-L26

(the source being unit in ex01).

The penultimate term is a superficial source; an example can be found in:

https://github.com/kinnala/scikit-fem/blob/6159aa309cd182578405ee8a84553d9a6d7462de/docs/examples/ex28.py#L102

(again for a unit load, but the `LinearForm` can be modified for nonuniform sources).

**Reply by abhishek-appana** [reply]
Hello @gdmcbain,

Thank you for your response! That wasn't my question.

I understand how to express them as linear/bilinear forms and assemble the terms. My question was in terms of solving the problem.

Below is the code I have written for this. It is not working. I am not sure what is wrong!

```python
@LinearForm
def tabs(v, _):
    return dt*q_tabs*v

@LinearForm
def loss(v, _):
    return dt*Q_gen*v

M = rho * C_p * asm(mass,basis['cell'])
L = k * asm(laplace, basis['cell'])
rhs_1 = -1 * (asm(tabs,basis['neg_tab']) + asm(tabs, basis['pos_tab']))
rhs_2 = asm(loss, basis['cell'])

T_init = np.repeat(298, basis['cell'].doflocs.shape[1])

A = M + 0.5 * L * dt
B = M - 0.5 * L * dt

backsolve = splu(A.T).solve

def evolve(t: float, 
           u: np.ndarray) -> Iterator[Tuple[float, np.ndarray]]:
    
    while t < 10:
        t, u = t + dt, backsolve((B @ u) + rhs_1 + rhs_2)
        yield t, u


if __name__ == '__main__':

    from argparse import ArgumentParser
    from pathlib import Path

    from matplotlib.animation import FuncAnimation
    import matplotlib.pyplot as plt

    from skfem.visuals.matplotlib import plot

    parser = ArgumentParser(description='heat equation in a rectangle')
    parser.add_argument('-g', '--gif', action='store_true', 
                        help='write animated GIF', )
    args = parser.parse_args()

    ax = plot(mesh, T_init[basis['cell'].nodal_dofs.flatten()], shading='gouraud')
    title = ax.set_title('t = 0.00')
    field = ax.get_children()[0]  # vertex-based temperature-colour
    fig = ax.get_figure()
    fig.colorbar(field)


    def update(event):
        t, u = event

        title.set_text(f'$t$ = {t:.2f}')
        field.set_array(u[basis['cell'].nodal_dofs.flatten()])

    animation = FuncAnimation(
        fig,
        update,
        evolve(0., T_init),
        repeat=False,
        interval=50,
    )
    if args.gif:
        animation.save(Path(__file__).with_suffix('.gif'), 'imagemagick')
    else:
        plt.show()
```

**Reply by gdmcbain** [reply]
Right, yes, sorry.  But when you say "it is not working", how do you mean?  What happens?  I can't run this here because a number of terms are undefined (`rho`, `C_p`, `basis`, &hellip;).

**Reply by abhishek-appana** [reply]
The result is displayed for t=0. But it is not getting solved post that.

Sorry about that. Here is the full code. It's slightly different from above (Had to add another rhs term)

```python
from typing import Iterator, Tuple

import numpy as np
from scipy.sparse.linalg import splu

from skfem import *
from skfem.helpers import dot, grad

length = 0.5
height = 0.5
ncells = 50

T_inf = 298
Q_gen = 1000
q_tabs = 500
rho = 25
C_p = 90
k = 25
h = 10

dt = 0.01

mesh = (MeshQuad
        .init_tensor(np.linspace(0, length, ncells),
                     np.linspace(0, height, ncells))
        .with_boundaries({'1': lambda x: np.isclose(x[1],0),
                          '2': lambda x: np.isclose(x[0],0),
                          '3': lambda x: np.isclose(x[1],0.5) & np.greater_equal(x[0],0) & np.less_equal(x[0],0.1),
                          '4': lambda x: np.isclose(x[1],0.5) & np.greater_equal(x[0],0.1) & np.less_equal(x[0],0.2),
                          '5': lambda x: np.isclose(x[1],0.5) & np.greater_equal(x[0],0.2) & np.less_equal(x[0],0.3),
                          '6': lambda x: np.isclose(x[1],0.5) & np.greater_equal(x[0],0.3) & np.less_equal(x[0],0.4),
                          '7': lambda x: np.isclose(x[1],0.5) & np.greater_equal(x[0],0.4) & np.less_equal(x[0],0.5),
                          '8': lambda x: np.isclose(x[0],0.5)}))

e = ElementQuad2()  # or ElementQuad1
basis = {
    'cell': Basis(mesh, e),
    'pos_tab': FacetBasis(mesh, e, facets=mesh.boundaries['4']),
    'neg_tab': FacetBasis(mesh, e, facets=mesh.boundaries['6']),
    'bottom': FacetBasis(mesh, e, facets=mesh.boundaries['1']),
    'left': FacetBasis(mesh, e, facets=mesh.boundaries['2']),
    'top_1': FacetBasis(mesh, e, facets=mesh.boundaries['3']),
    'top_2': FacetBasis(mesh, e, facets=mesh.boundaries['5']),
    'top_3': FacetBasis(mesh, e, facets=mesh.boundaries['7']),
    'right': FacetBasis(mesh, e, facets=mesh.boundaries['8'])}

@BilinearForm
def laplace(u, v, _):
    return dot(grad(u),grad(v))

@BilinearForm
def mass(u, v, _):
    return u*v

@LinearForm
def tabs(v, _):
    return dt*q_tabs*v

@BilinearForm
def rest(u, v, _):
    return dt*h*(u - T_inf)*v

@LinearForm
def loss(v, _):
    return dt*Q_gen*v

T_init = np.repeat(298, basis['cell'].doflocs.shape[1])

M = rho * C_p * asm(mass,basis['cell'])
L = k * asm(laplace, basis['cell'])
rhs_1 = -1 * (asm(tabs,basis['neg_tab']) + asm(tabs, basis['pos_tab']))
rhs_2 = (-1 * (asm(rest,basis['bottom']) + asm(rest,basis['left']) + asm(rest,basis['top_1']) 
              + asm(rest,basis['top_2']) + asm(rest,basis['top_3']) + asm(rest,basis['right'])))
rhs_3 = asm(loss, basis['cell'])

A = M + 0.5 * L * dt
B = M - 0.5 * L * dt

backsolve = splu(A.T).solve

def evolve(t: float, 
           u: np.ndarray) -> Iterator[Tuple[float, np.ndarray]]:
    
    while t < 10:
        t, u = t + dt, backsolve((B @ u) + rhs_1 + rhs_2 + rhs_3)
        yield t, u


if __name__ == '__main__':

    from argparse import ArgumentParser
    from pathlib import Path

    from matplotlib.animation import FuncAnimation
    import matplotlib.pyplot as plt

    from skfem.visuals.matplotlib import plot

    parser = ArgumentParser(description='heat equation in a rectangle')
    parser.add_argument('-g', '--gif', action='store_true', 
                        help='write animated GIF', )
    args = parser.parse_args()

    ax = plot(mesh, T_init[basis['cell'].nodal_dofs.flatten()], shading='gouraud')
    title = ax.set_title('t = 0.00')
    field = ax.get_children()[0]  # vertex-based temperature-colour
    fig = ax.get_figure()
    fig.colorbar(field)


    def update(event):
        t, u = event

        title.set_text(f'$t$ = {t:.2f}')
        field.set_array(u[basis['cell'].nodal_dofs.flatten()])

    animation = FuncAnimation(
        fig,
        update,
        evolve(0., T_init),
        repeat=False,
        interval=50,
    )
    if args.gif:
        animation.save(Path(__file__).with_suffix('.gif'), 'imagemagick')
    else:
        plt.show()
```

**Reply by abhishek-appana** [reply]
I will start a new discussion about this! Helps track it better
